docs/proposals/decoupling-triggers-and-runtimes.md

```diff
@@ -0,0 +1,117 @@
+     - "event"/"input": This is a blob in which the function receives the information to process. The blob could contain any structure and that will depend on the event source.
+     - "context": General information about the function environment. E.g Invoke ID, function version, function ARN…
+ - OpenWhisk: 
+   - Protocol: A Kafka service transforms HTTP Requests to Kafka messages in any case.
+   - Parameters: Functions receive a single argument "parameters" that contains a blob with the body of the HTTP request.
+ - Fission:
+   - Protocol: Functions request are received in the runtimes as HTTP request what gives the opportunity to give a response directly.
+   - Parameters: All the functions receive an object "context". This object has different properties depending on the runtime but as minimum it has a property "request" to read inputs and "response" to answer them.
+
+Regarding the available solutions and the current architecture of Kubeless I would choose a similar solution to Fission/OpenWhisk: Use the HTTP protocol to communicate runtimes and triggers and expose a single parameter. We can tweak this parameter depending on the runtime language in order to give different functionalities (if needed). This is the simplest solution for our use case, the most flexible and it is easy to use. 
+As specific proposal, the parameter should contain at least the information about the request. The properties of the request object will change depending on the runtime and the trigger source but in any case it should contain the parameter body with the inputs of the request (the body of a HTTP POST or the message of a Kafka entry). The body can be plain text or a JSON object, see the next section for specific details.
+
+### Function input
+Following the above premises and the [CNCF suggestion](https://docs.google.com/document/d/1UjW8bt5O8QBgQRILJVKZJej_IuNnxl20AJu9wA8wcdI/edit#heading=h.3s49zyc) this can be the a possible implementation for the input object that functions will receive the following schema (represented in JSON but the serialization may vary depending on the language):
+
+```json
+{
+  "data": "string|object",
+  "metadata": {
+    "type": "string",
+    "body": "string",
+    ["content-type": "string"],
+    ["path": "string"],
+    ["method": "string"],
+    ["headers": "object"],
+    ["topic": "string"],
+    ...
+  },
+  "context": {
+    "function-name": "string",
+    "runtime": "string",
+    "namespace": "string",
+    "memory-limit": "string",
+    ["schedule": "string"],
+    ["logger": "object"],
+    ...
+  }
+}
+```
+
+Note: Properties with brackets can be empty. Any required property can be added in the future maintaining backwards compatibility.
+
+ - Data: Serialized version of the request body
+ - Metadata: Information about the request
+   - Type: Type of the request ("HTTP", "PubSub", "Scheduled", ...)
```

Comment:
I propose to adopt the 4 invocation types that were mentioned in the whitepaper of serverless-wg of CNCF:
- synchronous
- asynchronous
- streaming
- batch

E.g. HTTP could be both. Just think of a web hook. In the asynchronous HTTP case, the response could be just a tracking ID and not the result of the function.

Reply:
I strongly believe triggers should be dumb, and cleverness should be in (composable) functions.

So for example your async HTTP example is just a _sync_ HTTP trigger+function that puts an event on a queue and then returns a tracking ID (again, synchronously).  This pattern can be used to turn any sync request into async (and you can think of the opposite version that blocks the caller until some sort of "async" response occurs).

I suggest *triggers* are inherently sync or async (or other properties you mention).  The differences are the way the *function* is invoked by that trigger - specifically, if it's worth generating a response object.  In this light, _all_ functions could be invoked synchronously, and we could just document that (eg) pubsub triggers discard any return value.

More interesting to me is "at most once" vs "at least once".  In particular: should the caller (ie: trigger) re-attempt calling the function several times until it gets a response, or should it "fire and forget" and return no response (not the usual interpretation of at-most-once, but useful in this context).  The at-most-once/at-least-once concept can be applied obviously to pubsub queues (using the same terminology), _and_ many types of HTTP notification/monitoring webhooks when we don't care about the response (even failure) (this is sometimes called level-triggered vs edge-triggered).

Reply:
Let's omit the `type` key, we can substitute it with a `source` so functions can adapt the logic, if necessary, depending on the source.

Regarding sync/async I agree to make functions always "sync". The returned value of the function can be "stringified" and sent back to the caller. That way functions can return function final results or a track ID.

Finally regarding the behavior, let's go with "at most once" since it is more similar to the way we are doing it today. If in the future we want to add the possibility of configuring a trigger with the "at least once" policy we will be able to do it.